### PR TITLE
Validation

### PR DESCRIPTION
**Pull request by kujawa98**

### CODE REVIEW

File: src/main/java/com/example/search/engine/exception/ValidationException.java [hunk 1]
```diff
@@ -0,0 +1,7 @@
+package com.example.search.engine.exception;
+
+public class ValidationException extends RuntimeException{
```
Comment (pawel-ozog) [inline src/main/java/com/example/search/engine/exception/ValidationException.java]: Please rename this class to better reflect its purpose, like: InvalidInputValidationException

File: src/main/java/com/example/search/engine/service/ValidationService.java [hunk 2]
```diff
@@ -0,0 +1,24 @@
...
+import org.springframework.util.Assert;
+
+import java.util.List;
+
+
+public class ValidationService implements Validator {
+
+    public void validate(List<String> input) {
+        Assert.notNull(input,"input can not be null");
+        String illegalSymbols = "!@#$%^&*()_+-=";
```
Comment (pawel-ozog) [inline src/main/java/com/example/search/engine/service/ValidationService.java]: Please refactor illegalSymbols and make it private static final.

File: src/main/java/com/example/search/engine/service/ValidationService.java [hunk 3]
```diff
@@ -0,0 +1,24 @@
...
+        input.stream()
+                .filter(word -> StringUtils.containsAny(word, illegalSymbols))
+                .findFirst()
+                .ifPresent(s -> {
+                    throw new ValidationException("Illegal argument appeared");
```
Comment (pawel-ozog) [inline src/main/java/com/example/search/engine/service/ValidationService.java]: Please add invalid word into the exception's message

File: src/main/java/com/example/search/engine/Runner.java [hunk 4]
```diff
@@ -26,14 +28,17 @@ public void run(String... args) throws ClassNotFoundException, NoSuchMethodExcep
...
 
         Try.of(() -> new ContentReadService(FileContentSupplier.of(inputResource)).readLinesFromFile())
                 .peek(input -> log.info("Read input size {}", input.size()))
+                .peek(validator::validate)
                 .map(LongWordService::new)
                 .map(LongWordService::prepareLongWord)
                 .peek(longWord -> log.info("Long Word size {}, Prepared milestones {}", longWord.content().length(), longWord.milestones()))
                 .map(longWord -> Tuple.of(longWord, new ContentReadService(FileContentSupplier.of(searchResource)).readLinesFromFile()))
                 .peek(tuple -> log.info("To match {}", tuple._2()))
+                .peek(tuple -> validator.validate(tuple._2))
```
Comment (pawel-ozog) [inline src/main/java/com/example/search/engine/Runner.java:41]: Please add additional 'onFailure' section and log out validation error

File: src/test/java/com/example/search/engine/service/ValidationServiceTest.java [hunk 5]
```diff
@@ -0,0 +1,36 @@
...
+import java.util.List;
+
+import static org.junit.jupiter.api.Assertions.*;
+
+public class ValidationServiceTest {
+    private final ValidationService validationService = new ValidationService();
+
+    @Test
+    void testValidationException() {
+        List<String> words = Arrays.asList("firstWord", "secondWord", "thi@dW&rd");
```
Comment (pawel-ozog) [inline src/test/java/com/example/search/engine/service/ValidationServiceTest.java]: How about an array having invalid word at index[0]?

File: src/main/java/com/example/search/engine/helper/MilestoneResolver.java [hunk 6]
```diff
@@ -10,14 +10,18 @@
...
 
+    /**
+     * Key is a word's length
+     * Value is a index of first word whose lenght's equal to key
+     * */
     private final SortedMap<Integer, Integer> milestones;
 
     private MilestoneResolver(final SortedMap<Integer, Integer> milestones) {
-        Assert.notNull(milestones, "milestones must not be null");
+        Assert.notNull(milestones, "milestones must not be null"); //throws exception if milestone is null
```
Comment (pawel-ozog) [inline src/main/java/com/example/search/engine/helper/MilestoneResolver.java]: This comment is redundant. Please delete it.

File: src/main/java/com/example/search/engine/helper/MilestoneResolver.java [hunk 7]
```diff
@@ -10,14 +10,18 @@
...
         this.milestones = milestones;
     }
 
-    public static MilestoneResolver of(SortedMap<Integer, Integer> milestones) {
+    public static MilestoneResolver of(SortedMap<Integer, Integer> milestones) { //of is a factory method
```
Comment (pawel-ozog) [inline src/main/java/com/example/search/engine/helper/MilestoneResolver.java]: This comment is redundant. Please delete it.

**Review by pawel-ozog — CHANGES_REQUESTED** [review]

File: src/main/java/com/example/search/engine/service/ValidationService.java [hunk 8]
```diff
@@ -0,0 +1,24 @@
...
+
+import java.util.List;
+
+
+public class ValidationService implements Validator {
+
+    public void validate(List<String> input) {
+        Assert.notNull(input,"input can not be null");
+        String illegalSymbols = "!@#$%^&*()_+-=";
+        input.stream()
```
Comment (pawel-ozog) [inline src/main/java/com/example/search/engine/service/ValidationService.java]: We expect a very large number of inputs and here we do not care of the order. To increase performance please use parallelStream

**Review by pawel-ozog — CHANGES_REQUESTED** [review]

File: src/main/java/com/example/search/engine/service/ValidationService.java (hunk 2 above)
Comment (kujawa98) [reply src/main/java/com/example/search/engine/service/ValidationService.java]: What do you mean by refactor?

File: src/test/java/com/example/search/engine/service/ValidationServiceTest.java [hunk 9]
```diff
@@ -0,0 +1,40 @@
...
+import java.util.Collections;
+import java.util.List;
+
+import static org.junit.jupiter.api.Assertions.*;
+
+public class ValidationServiceTest {
+    private final ValidationService validationService = new ValidationService();
+
+    @Test
+    void testValidationException() {
```
Comment (pawel-ozog) [inline src/test/java/com/example/search/engine/service/ValidationServiceTest.java]: Please use @ParametrizedTest to perform multiple assertions with one test

**Review by pawel-ozog — APPROVED** [review]

**Pull request merged**